File: apps/java/src/main/java/chapter08/DummyEncoder.java
package chapter08;

import chapter02.Row;
import chapter02.Table;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** カテゴリ値の列を、最初のカテゴリを除いたカテゴリごとの 0 と 1 の列（ダミー変数）にする前処理。 */
public record DummyEncoder(List<String> columns) implements Transformer {
  public DummyEncoder {
    columns = List.copyOf(columns);
  }

  @Override
  public FittedTransformer fit(Table x) {
    return new FittedDummyEncoder(dummiesOf(x));
  }

  /** 列ごとに、ダミー変数にするカテゴリ（最初のカテゴリを除く）を求める。 */
  private Map<String, List<String>> dummiesOf(Table x) {
    Map<String, List<String>> dummies = new LinkedHashMap<>();
    for (String column : columns) {
      List<String> categories = categoriesOf(x, column);
      dummies.put(column, categories.subList(1, categories.size()));
    }
    return dummies;
  }

  /** 列の値を重複なく並べ替える。欠損値は除く。 */
  private static List<String> categoriesOf(Table x, String column) {
    return x.rows().stream()
        .filter(row -> !row.isMissing(column))
        .map(row -> row.text(column))
        .distinct()
        .sorted()
        .toList();
  }

  /** fit で求めたカテゴリを持ち、どのデータにも同じダミー変数の列を作る。 */
  public record FittedDummyEncoder(Map<String, List<String>> dummies) implements FittedTransformer {
    public FittedDummyEncoder {
      // Map.copyOf は順序を保たないので、列の順を保つ LinkedHashMap に写す。
      // subList が返す部分リストはシリアライズできないので、List.copyOf で写す
      Map<String, List<String>> copy = new LinkedHashMap<>();
      dummies.forEach((column, categories) -> copy.put(column, List.copyOf(categories)));
      dummies = Collections.unmodifiableMap(copy);
    }

    @Override
    public Table transform(Table x) {
      return encode(x, dummies);
    }
  }

  /** 元の列を除き、ダミー変数の列を末尾に足す。 */
  static Table encode(Table x, Map<String, List<String>> dummies) {
    List<String> columns = new ArrayList<>(x.columns());
    dummies.forEach(
        (column, categories) -> {
          columns.remove(column);
          categories.forEach(category -> columns.add(column + "_" + category));
        });
    List<Row> rows = x.rows().stream().map(row -> encode(row, dummies)).toList();
    return new Table(columns, rows);
  }

  private static Row encode(Row row, Map<String, List<String>> dummies) {
    Row encoded = row;
    for (Map.Entry<String, List<String>> entry : dummies.entrySet()) {
      String value = row.text(entry.getKey());
      for (String category : entry.getValue()) {
        encoded =
            Rows.with(encoded, entry.getKey() + "_" + category, value.equals(category) ? "1" : "0");
      }
    }
    return encoded;
  }
}

File: apps/csharp/src/MachineLearning/Chapter08/DummyEncoder.cs
namespace MachineLearning.Chapter08;

using MachineLearning.Chapter02;

/// <summary>カテゴリ値の列を、最初のカテゴリを除いたカテゴリごとの 0 と 1 の列（ダミー変数）にする前処理。</summary>
public sealed record DummyEncoder(IReadOnlyList<string> Columns) : ITransformer
{
    public IFittedTransformer Fit(Table x)
    {
        ArgumentNullException.ThrowIfNull(x);
        return new Fitted(
            [.. this.Columns.Select(column => new Dummies(column, [.. CategoriesOf(x, column).Skip(1)]))]);
    }

    /// <summary>列の値を重複なく並べ替える。欠損値は除く。</summary>
    private static IReadOnlyList<string> CategoriesOf(Table x, string column) =>
        [.. x.Rows
            .Where(row => !row.IsMissing(column))
            .Select(row => row.Text(column))
            .Distinct(StringComparer.Ordinal)
            .Order(StringComparer.Ordinal)];

    /// <summary>1 つの列と、その列でダミー変数にするカテゴリ（最初のカテゴリを除く）。</summary>
    public sealed record Dummies(string Column, IReadOnlyList<string> Categories);

    /// <summary>Fit で求めたカテゴリを持ち、どのデータにも同じダミー変数の列を作る。</summary>
    public sealed record Fitted(IReadOnlyList<Dummies> Dummies) : IFittedTransformer
    {
        public Table Transform(Table x)
        {
            ArgumentNullException.ThrowIfNull(x);
            var columns = x.Columns.ToList();
            foreach (var dummies in this.Dummies)
            {
                columns.Remove(dummies.Column);
                columns.AddRange(dummies.Categories.Select(category => $"{dummies.Column}_{category}"));
            }

            return new Table(columns, [.. x.Rows.Select(row => this.Encode(x.Columns, row))]);
        }

        private Row Encode(IReadOnlyList<string> columns, Row row)
        {
            // 元の列のセルを写してから、ダミー変数の列を足す。元のカテゴリの列は表の列から外れるので残っていてよい
            var cells = columns.ToDictionary(name => name, row.Text, StringComparer.Ordinal);
            foreach (var dummies in this.Dummies)
            {
                var value = row.Text(dummies.Column);
                foreach (var category in dummies.Categories)
                {
                    cells[$"{dummies.Column}_{category}"] =
                        string.Equals(value, category, StringComparison.Ordinal) ? "1" : "0";
                }
            }

            return new Row(cells);
        }
    }
}
